File: src/lib/__tests__/session-idle.test.ts
import { describe, expect, it } from "vitest";
import fixture from "./fixtures/session-idle-scenarios.json";
import {
  SESSION_IDLE_NO_FIX_S,
  SESSION_IDLE_STATIONARY_S,
  SESSION_PROGRESS_EPSILON_M,
  sessionIdleStep,
  type SessionIdleInput,
} from "../session-idle";
import {
  PRESUMED_ARRIVAL_CAR,
  PRESUMED_ARRIVAL_WALK,
  PROGRESS_EPSILON_M,
} from "../final-approach";

describe("sessionIdleStep (공유 fixture)", () => {
  for (const s of fixture.scenarios) {
    it(s.name, () => {
      expect(sessionIdleStep(s.input as SessionIdleInput)).toBe(s.expect);
    });
  }

  it("무효 입력(음수·NaN·무한)은 null", () => {
    expect(sessionIdleStep({ secondsSinceUsableFix: -1, secondsSinceProgress: 0 })).toBeNull();
    expect(sessionIdleStep({ secondsSinceUsableFix: NaN, secondsSinceProgress: 0 })).toBeNull();
    expect(sessionIdleStep({ secondsSinceUsableFix: 0, secondsSinceProgress: Infinity })).toBeNull();
  });

  it("국면 무관 안전망은 도착 추정보다 모든 축이 느슨하다(両프로파일 — 경로 중간 정상 이동을 끊지 않는 조건)", () => {
    for (const p of [PRESUMED_ARRIVAL_WALK, PRESUMED_ARRIVAL_CAR]) {
      expect(SESSION_IDLE_NO_FIX_S).toBeGreaterThan(p.noFixSeconds);
      expect(SESSION_IDLE_STATIONARY_S).toBeGreaterThan(p.stationarySeconds);
    }
    expect(SESSION_PROGRESS_EPSILON_M).toBeGreaterThan(PROGRESS_EPSILON_M);
  });

  it("무이동 축이 없으면(null) 두절 축만 산다", () => {
    expect(sessionIdleStep({ secondsSinceUsableFix: 600, secondsSinceProgress: null })).toBe("noFix");
    expect(sessionIdleStep({ secondsSinceUsableFix: NaN, secondsSinceProgress: null })).toBeNull();
  });
});

File: android/kit/src/test/kotlin/space/dodoplanet/gildongmu/kit/SessionIdleTest.kt
package space.dodoplanet.gildongmu.kit

import kotlinx.serialization.Serializable
import kotlin.test.Test
import kotlin.test.assertEquals
import kotlin.test.assertNull
import kotlin.test.assertTrue

/** 웹 정본과 같은 공유 fixture(`session-idle-scenarios.json`)를 읽는다 — Kit `SessionIdleTests` 미러. */
class SessionIdleTest {
    @Serializable
    private data class FixtureFile(val scenarios: List<Scenario>) {
        @Serializable
        data class Scenario(val name: String, val input: Input, val expect: String? = null)

        @Serializable
        data class Input(val secondsSinceUsableFix: Double, val secondsSinceProgress: Double? = null)
    }

    @Test fun `공유 fixture 동조`() {
        val scenarios = Fixtures.sharedJson("session-idle-scenarios.json", FixtureFile.serializer()).scenarios
        assertTrue(scenarios.isNotEmpty())
        for (s in scenarios) {
            val got = sessionIdleStep(s.input.secondsSinceUsableFix, s.input.secondsSinceProgress)
            assertEquals(s.expect, got?.rawValue, s.name)
        }
    }

    @Test fun `무효 입력은 null`() {
        assertNull(sessionIdleStep(-1.0, 0.0))
        assertNull(sessionIdleStep(Double.NaN, 0.0))
        assertNull(sessionIdleStep(0.0, Double.POSITIVE_INFINITY))
    }

    @Test fun `국면 무관 안전망은 도착 추정보다 모든 축이 느슨하다 — 두 프로파일`() {
        for (p in listOf(PresumedArrivalThresholds.walk, PresumedArrivalThresholds.car)) {
            assertTrue(sessionIdleNoFixSeconds > p.noFixSeconds)
            assertTrue(sessionIdleStationarySeconds > p.stationarySeconds)
        }
        assertTrue(sessionProgressEpsilonMeters > progressEpsilonMeters)
    }

    @Test fun `무이동 축이 없으면 두절 축만 산다`() {
        assertEquals(SessionIdleReason.noFix, sessionIdleStep(600.0, null))
        assertNull(sessionIdleStep(Double.NaN, null))
    }
}
